[12.0][NEW] crm_task module

v-garcies opened the pull request:
I developed this module which links project tasks to CRM opportunities.

This is my first contribution, can you guide me if there are changes to be made?

v-garcies committed:
[12.0][ADD] crm_task : this new module helps users to link tasks to CRM opportunities

v-garcies committed:
[12.0][ADD] crm_task : this new module helps users to link tasks to CRM opportunities

v-garcies commented:
Do I need to do something more?

rousseldenis commented:
@v-garcies Please check https://runbot3.odoo-community.org/runbot/static/build/3425819-336-cef348/logs/job_20_test_all.txt warnings

HaraldPanten commented:
Hi @v-garcies are you going to attend @rousseldenis comments?

THX

pedrobaeza commented:
What is the difference with https://github.com/odoo/odoo/tree/12.0/addons/crm_project ?

v-garcies commented:
@pedrobaeza This module adds to the CRM opportunities a smart button with the number and direct access to the related tasks

pedrobaeza commented:
Yeah, but it adds no value IMO to the existing module.

v-garcies commented:
The goal is to be able to find the tasks related to an opportunity, with the crm_project module this is impossible ... Maybe we should add this improvement to the crm_project module directly ?

pedrobaeza commented:
If I don't remember bad, in the chatter an entry is created when linked to the task (that should mostly be only one, not several).

v-garcies commented:
I don't see information about the associated opportunities on the task form. In addition, crm_project acts only on the leads and not on the opportunities.

This module has a different purpose, it allows you to follow the tasks carried out from the opportunity.

Example of opportunity without task :
<img width="1792" alt="opportunity-form_without related-task" src="https://user-images.githubusercontent.com/19540535/86621780-14fcec00-bfbf-11ea-9b05-32692f85c207.png">

Example of opportunity with 2 related tasks :
<img width="1792" alt="opportunity-form" src="https://user-images.githubusercontent.com/19540535/86621666-da934f00-bfbe-11ea-8e75-846262f50f20.png">

Example of list of linked tasks by clicking on the smart button :
<img width="1792" alt="related-tasks-list" src="https://user-images.githubusercontent.com/19540535/86621677-e1ba5d00-bfbe-11ea-8ddd-07e4eb6315e9.png">

pedrobaeza commented:
OK, let's see if reviewers find it useful.

HaraldPanten reviewed (approved):
It's functionally working fine. LGTM 👍 

@ValentinVinagre

ValentinVinagre commented on crm_task/models/crm_lead.py:
```diff
@@ -0,0 +1,20 @@
+        string="Task", compute='_compute_task_count',
+        required=False)
+
+    @api.depends('partner_id')
```
Why does it depend on the partner_id if the customer_id isn't used in the '_compute_task_count' function?
Could the taskconunt field  be calculated like this?
```
def _compute_task_count(self):
    self.taskcount = len(self. task_ids)
```

v-garcies replied:
I just made the changes you suggest

ValentinVinagre commented on crm_task/models/project_task.py:
```diff
@@ -0,0 +1,10 @@
+class LeadinTask(models.Model):
+    _inherit = 'project.task'
+
+    lead_id = fields.Many2one("crm.lead", string="Opportunity", required=False)
```
why the required false? by default is false.

v-garcies replied:
I just deleted that, thanks for your feedback

ValentinVinagre reviewed (commented):
Some comments.

v-garcies commented:
The Travis CI build failed due to crm_phonecall_planner module :

crm_phonecall_planner/wizards/crm_phonecall_planner.py:171: [E8103(sql-injection), CrmPhonecallPlan.action_accept] SQL injection risk. Use parameters if you can. - More info https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#no-sql-injection

pedrobaeza commented:
That are not the real errors, but warnings. Yours are:
```
crm_task/models/crm_lead.py:4:1: F401 'odoo.api' imported but unused
crm_task/models/crm_lead.py:18:5: E303 too many blank lines (2)
```

v-garcies commented:
> That are not the real errors, but warnings. Yours are:
> 
> ```
> crm_task/models/crm_lead.py:4:1: F401 'odoo.api' imported but unused
> crm_task/models/crm_lead.py:18:5: E303 too many blank lines (2)
> ```

Ok sorry, I have a little trouble reading the travis report ...
So, I have corrected the points listed 👍

ValentinVinagre commented on crm_task/models/crm_lead.py:
```diff
@@ -0,0 +1,17 @@
+
+    taskcount = fields.Integer(
+        string="Task", compute='_compute_task_count',
+        required=False)
```
the field is required=False by default, it's not necessary.

v-garcies replied:
ok, I will remove this attribute 👍

ValentinVinagre commented on crm_task/views/crm_lead.xml:
```diff
@@ -0,0 +1,30 @@
+        <field name="arch" type="xml">
+            
+            <!-- Smart button - Tasks -->
+            <xpath expr="//sheet/div[@name='button_box']" position="inside">
+                <button class="oe_stat_button" type="action" id="crm_lead_button_task" 
+                    name="%(crm_task_to_opportunity)d" icon="fa-tasks"
+                    context="{'search_default_lead_id': [id], 'default_name': name, 'default_description': description}">
+                    <field string="Tasks" name="taskcount" widget="statinfo"/>
+                </button>
+            </xpath>
+
+            <!-- Page - Tasks list -->
+            <xpath expr="//form/sheet/notebook/page[1]" position="before">
+                <page string="Tasks" name="tasks" attrs="{'invisible':[('taskcount', '=', 0)]}">
+                    <field name="task_ids"/>
+                </page>
+            </xpath>
```
Why do you show the task twice? Isn't enough with one field?

v-garcies replied:
My clients regularly tell me that odoo is great but that the majority of things are too long to perform or require too many clicks.
So I try to improve the user experience by making the information more readable without necessarily having to change views.

However the smart button is still necessary if you want to consult the tasks in calendar view type for example

Do you have a better idea?

ValentinVinagre replied:
The main problem is that with the two fields the information is duplicated, so it should only be shown 1 way, IMO the button is the best option.

ValentinVinagre commented on crm_task/views/project_task.xml:
```diff
@@ -0,0 +1,66 @@
+    </record> 
+
+    <!-- ACTION -->
+    <record id="crm_task_to_opportunity" model="ir.actions.act_window">
```
Why do you create a new action? Can't you use one of the existing?

v-garcies replied:
I created a new action to remove the automatic filter "My task" and especially to open the tasks in list view by default

If I use one of the existing actions we come across this view by clicking on the task button from an opportunity with no related tasks :
<img width="1494" alt="Capture d’écran 2020-07-09 à 12 36 07" src="https://user-images.githubusercontent.com/19540535/87029841-fe35df80-c1e0-11ea-98f2-43a22b9a939f.png">

ValentinVinagre replied:
Okay. A little more, I would move the "opportunity" field down to "Assigned to" in the task view.
![Captura de pantalla 2020-07-14 a las 12 33 25](https://user-images.githubusercontent.com/22913794/87416211-aa573c00-c5ce-11ea-8a54-664408182649.png)

ValentinVinagre reviewed (commented):
Some comments.

HaraldPanten commented:
@v-garcies please, squash commits

v-garcies commented:
> @v-garcies please, squash commits

Hi, ok how should I do this ? 🤔

HaraldPanten commented:
https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history#reword-other-commit-messages-take-2

git rebase -i HEAD~x ( x: depending on the number of your last commits that you need to squash)

In this example, you have some "administrative" commits, like:

<img width="1352" alt="Captura de pantalla 2020-07-21 a las 9 33 37" src="https://user-images.githubusercontent.com/32449357/88025963-6667b800-cb35-11ea-85c3-b099576dd4fe.png">

You should squash them and only keep the commits that make sense like the [ADD] ones, [IMP] ones, important [FIX] ones... 

In you example, keeping commits where you say that you're fixing travis/runbot errors, or where you say that you're taking in consideration ValentinVinagre's comments, doesn't make sense.

Do you know what I mean?

v-garcies commented:
> https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history#reword-other-commit-messages-take-2
> 
> git rebase -i HEAD~x ( x: depending on the number of your last commits that you need to squash)
> 
> In this example, you have some "administrative" commits, like:
> 
> <img alt="Captura de pantalla 2020-07-21 a las 9 33 37" width="1352" src="https://user-images.githubusercontent.com/32449357/88025963-6667b800-cb35-11ea-85c3-b099576dd4fe.png">
> 
> You should squash them and only keep the commits that make sense like the [ADD] ones, [IMP] ones, important [FIX] ones...
> 
> In you example, keeping commits where you say that you're fixing travis/runbot errors, or where you say that you're taking in consideration ValentinVinagre's comments, doesn't make sense.
> 
> Do you know what I mean?

Ok, thanks for the explanations, I'll take care of that

HaraldPanten commented:
Hi @v-garcies ,

As your module is an [ADD] you should leave only the first commit because you only have 1 relevant commit. Look at this example from another PR:

<img width="1195" alt="Captura de pantalla 2020-07-21 a las 11 05 59" src="https://user-images.githubusercontent.com/32449357/88035713-da5c8d00-cb42-11ea-9706-b70b2aaa241c.png">

So, regarding your commits, you should leave only the first one and you should rename this one following the structure --> [VERSION][ADD] module_name =>(normally) [12.0][ADD] crm_task or something like [12.0][ADD] new module crm_task.
Also a small description of the commit will be welcome; for example: this new module helps users to link tasks to CRM opportunities.

<img width="1431" alt="Captura de pantalla 2020-07-21 a las 11 07 03" src="https://user-images.githubusercontent.com/32449357/88035964-2c051780-cb43-11ea-89f8-401cf8e9eb3d.png">

v-garcies commented:
@HaraldPanten My first git squash went very badly, I deleted all the last commits even though I used the "fixup" option ... I just corrected all that right now. Thank you for your explanations!

Hope all is well this time :)

HaraldPanten commented:
Looks better now 👍 

Let's wait for @pedrobaeza , @ValentinVinagre and/or others reviews.

pedrobaeza commented:
Well, I don't see it needed, but if others do, then those should review.

ValentinVinagre commented on crm_task/readme/USAGE.rst:
```diff
@@ -0,0 +1,5 @@
+
+1. Go to Lead
+2. A new button allows you to create/access project tasks
+3. A new page list all related tasks from the opportunity
```
This feature doesn't exist. You could explain how to link a task with a lead/opportunity.

ValentinVinagre reviewed (changes requested):
A little change.

ValentinVinagre commented on crm_task/views/project_task.xml:
```diff
@@ -0,0 +1,66 @@
+    <record id="project_task_form2" model="ir.ui.view">
+        <field name="name">project.task.form.2</field>
+        <field name="model">project.task</field>
+        <field name="inherit_id" ref="hr_timesheet.view_task_form2_inherited"/>
```
As you're working on 'user_id' you don't need to use ref="hr_timesheet.view_task_form2_inherited" because this field is in "project.view_task_form2". Right?

ValentinVinagre commented on crm_task/views/project_task.xml:
```diff
@@ -0,0 +1,66 @@
+        <field name="model">project.task</field>
+        <field name="inherit_id" ref="project.view_task_search_form"/>
+        <field name="arch" type="xml">
+            
```
Please delete the empty lines

ValentinVinagre commented on crm_task/__manifest__.py:
```diff
@@ -0,0 +1,28 @@
+    'depends': [
+        'crm',
+        'project',
+        'hr_timesheet',
```
Remove dependency if you change the inherited view.

ValentinVinagre reviewed (commented):
The last changes :D the last effort.

v-garcies committed:
[12.0][ADD] crm_task : this new module helps users to link tasks to CRM opportunities

v-garcies committed:
Merge branch 'crm_task' of https://github.com/v-garcies/crm into crm_task

HaraldPanten commented:
Hi @v-garcies you should squash commits, as you have 4 different [ADD].

claudiagn reviewed (approved):
functional review ok!

dreispt commented:
/ocabot merge nobump

pedrobaeza commented:
Aborting, as this PR includes a merge commit that messes up the repo. @dreispt please don't merge it so lightly, as others then have to fix the errors this provokes.

dreispt commented:
Hello @pedrobaeza, merge decision is made based on the PR Reviews, and I see there no strong blocking reason there.
I see no changes that would break the repo, and the build is green. Can you elaborate please?

pedrobaeza commented:
Merge commits included in a pull request (like the fourth commit here), scrambles the commit history in the repo, having multiple parent ramifications, and when using the `git format-patch` method we used for rescuing commit history when migrating modules, this makes to crash and have conflicts, so always avoid such things. In fact, thanks to our bot that rebases the branch before merging, we are avoiding this problem more frequently.
